src/librustc/hir/lowering.rs

```diff
@@ -296,6 +296,14 @@ enum AnonymousLifetimeMode {
     PassThrough,
 }
 
+trait FnRetTransform {
+    fn lower(&self, ctx: &mut LoweringContext, output: &FunctionRetTy, impl_trait_context: ImplTraitContext) -> hir::FunctionRetTy;
+}
+
+struct NonAsyncRet;
+struct AsyncClosureRet;
+struct AsyncFunctionRet;
```

Comment:
I'm surprised you're not using an `enum` here.

Reply:
I watched a talk at a conference about how great dependency injection is right before I wrot this code :) (the talk was by Sandi Metz)

I don't mind switching to a method on an enum which then delegates to three other methods, but it introduces the opportunity for irregularity (adding an argument that only applies to 1 case, for example, which I think is usually suspicious). I'd love if Rust made it more natural to write this pattern.

Reply:
the main thing I don't want to do is pass an enum down and then write the match inside of this function. I think the proliferation of that pattern in lowering has made the code more difficult to understand (but I like OO, so YMMV).

Reply:
Our mileage does vary - `enum` represent finite state, whereas a trait introduces an open-world source of ambiguity - every single time there's a trait object involved, all the impls must be found to reason about anything at all.

I can's speak for all of @rust-lang/compiler, but I personally wouldn't use a `trait` unless the implementers are used in unrelated situations, but they can reuse some common abstraction through the `trait` (e.g. visitor/folder pattern) - in other words, bridging gaps in an open-world.

And even then I'd almost never employ a `trait` object (or `fn` pointers for that matter) except for bypassing polymorphic recursion (typically needed when a recursive function takes a closure) or when polymorphism is somehow prohibited (e.g. the query engine, which does dependency injection with `fn` pointers, or the codegen backend, which is dynamically loaded nowadays for LLVM reasons).

In this case, all the implementors are in the same location, they're not recursive data types (or any data at all, but most data wouldn't be relevant), they're created from similar `match`-es on `asyncness`, and the sole method of the trait has exactly one caller.

> I don't mind switching to a method on an enum which then delegates to three other methods, but it introduces the opportunity for irregularity (adding an argument that only applies to 1 case, for example, which I think is usually suspicious).

I don't understand this - are you referring to an argument that would come from a field of the `struct`s (which are currently all units)? Or from the caller of `lower`? Because in the latter case I don't know how that'd change anything, and in the former you just move all the `struct` fields to the respective `enum` variant.